src/main/java/org/folio/rest/impl/TenantPermsAPI.java

```diff
@@ -226,28 +211,79 @@ public void postTenantpermissions(OkapiPermissionSet entity, Map<String, String>
-              connection, vertxContext, tenantId)
-              .compose(v -> handleRenamedPerms(moduleId, getRenamedPerms(dbPerms, perms),
-                  connection, vertxContext, tenantId))
-              .compose(v -> handleRemovedPerms(getRemovedPerms(dbPerms, perms), connection));
+          return handleNewPerms(moduleId, dbPerms, perms, connection, vertxContext, tenantId)
+              .compose(x -> handleModifiedPerms(moduleId, dbPerms, perms, connection, vertxContext, tenantId))
+              .compose(v -> handleRenamedPerms(moduleId, dbPerms, perms, connection, vertxContext, tenantId))
+              .compose(v -> handleRemovedPerms(getRemovedPerms(dbPerms, perms), connection))
+              .compose(v -> migratePermsAssign(moduleId, dbPerms, connection));
         });
   }
 
-  private Future<Void> handleNewAndModifiedPerms(ModuleId moduleId, List<OkapiPermission> permList,
-      Conn connection, Context vertxContext, String tenantId) {
-    if (permList.isEmpty()) {
+  private Future<Void> migratePermsAssign(ModuleId moduleId, Map<String, Permission> dbPerms, Conn connection) {
+    // only if are upgrading from an earlier mod-permissions module do we upgrade users.
+    Permission permission = dbPerms.get(PermissionUtils.PERMS_USERS_ASSIGN_MUTABLE);
+    if (!"mod-permissions".equals(moduleId.getProduct())
+        || (permission != null && !permission.getDeprecated())) {
       return Future.succeededFuture();
     }
-    return savePermList(moduleId, permList, connection, vertxContext, tenantId);
+    CQLWrapper all = new CQLWrapper();
+    return connection.get(TABLE_NAME_PERMSUSERS, PermissionUser.class, all, false)
+        .compose(results -> {
+          Future<Void> future = Future.succeededFuture();
+          for (PermissionUser permUser : results.getResults()) {
+            JsonArray originalList = new JsonArray(permUser.getPermissions());
+            JsonArray newList = new JsonArray(permUser.getPermissions());
+            List<String> permissionsToBeAdded;
+            if (originalList.contains(PermissionUtils.PERMS_OKAPI_ALL)) {
+              permissionsToBeAdded = List.of(PermissionUtils.PERMS_USERS_ASSIGN_OKAPI,
+                  PermissionUtils.PERMS_USERS_ASSIGN_IMMUTABLE,
+                  PermissionUtils.PERMS_USERS_ASSIGN_MUTABLE);
+            } else if (originalList.contains(PermissionUtils.PERMS_PERMS_ALL)) {
+              permissionsToBeAdded = List.of(PermissionUtils.PERMS_USERS_ASSIGN_IMMUTABLE,
+                  PermissionUtils.PERMS_USERS_ASSIGN_MUTABLE);
+            } else {
+              permissionsToBeAdded = Collections.emptyList();
+            }
+            boolean added = false;
+            for (String perm : permissionsToBeAdded) {
+              if (!originalList.contains(perm)) {
+                newList.add(perm);
+                added = true;
+              }
+            }
+            if (added) {
```

Comment:
I prefer
`if (newList.size() > originalList.size()) {`

Reply:
That would work.. But I think current behavior is more clearer.